wallet/src/Pos/Wallet/Web/Methods/Payment.hs

```diff
@@ -214,3 +223,50 @@ sendMoney SendActions{..} passphrase moneySource dstDistr = do
 
     logDebug "sendMoney: constructing response"
     fst <$> constructCTx srcWallet srcWalletAddrs diff th
+
+
+-- | For each specified transaction, if it has been canceled, creates
+-- new transaction with correspondent destination address and money.
+-- This works as `mapM newPayment`, so completion may take a lot of time.
+reformCanceledTxs
+    :: MonadWalletWebMode m
+    => SendActions m -> PassPhrase -> ReformCanceledTxsParams -> m [CTx]
+reformCanceledTxs sendActions passphrase params = do
+    ptxs <- getPendingTxs
+    let canceledTxs :: [(TxId, (CId Wal, PtxPoolInfo))]
+        canceledTxs = flip mapMaybe ptxs $ \PendingTx{..} ->
+            case _ptxCond of
+                PtxWontApply _ poolInfo -> Just (_ptxTxId, (_ptxWallet, poolInfo))
+                _                       -> Nothing
+    logDebug $ sformat ("List of all canceled transactions ever: "%listJson)
+         (map fst canceledTxs)
+
+    txsToRecreate <- case rctpBanned params of
+        [] -> pure canceledTxs
+        bannedCTxs -> do
+            bannedTxs <- mapM decodeCTypeOrFail bannedCTxs
+            let bannedTxsSet = S.fromList bannedTxs
+            pure $ filter ((`S.notMember` bannedTxsSet) . fst) canceledTxs
+
+    logDebug $ sformat ("Transactions selected for recreation: "%listJson)
+        (map fst txsToRecreate)
+
+    fmap catMaybes . forM txsToRecreate $ \(txId, (wid, txEntry)) ->
```

Comment:
Again, this would be better written is a `foldlM` rather than doing a `forM` followed by another linear scan.

Reply:
Thanks, nice point!